src/fetch.rs

```diff
@@ -48,191 +47,144 @@ impl Method {
-pub fn fetch(
-    method: Method,
-    url: &str,
-    request_opts: Option<RequestOpts>,
-    callback: Box<Fn(web_sys::Response)>
-) {
-    let request = fetch_helper(method, url, request_opts);
-    let window = web_sys::window().expect("Can't find window");
-
-    // We could get the fetch_request_with_init promise directly from the func,
-    // but am running into issues with futures etc.
-    let request_promise = window.fetch_with_request(&request);
-    let f = wasm_bindgen_futures::JsFuture::from(request_promise)
-
-        .and_then(move |resp_value| {
-        // `resp_value` is a `Response` object.
-            assert!(resp_value.is_instance_of::<web_sys::Response>());
-        let resp: web_sys::Response = resp_value.dyn_into()
-            .expect("Problem casting response as Reponse.");
-
-        callback(resp);
-        future::ok(JsValue::null())
-    });
-
-    wasm_bindgen_futures::future_to_promise(f);
-}
+    pub fn fetch_string(self) -> impl Future<Item = String, Error = JsValue> {
+        self.fetch()
+            .and_then(|x| {
+                JsFuture::from(x.text().expect("Error retrieving text from a Response"))
+            })
+            // TODO handle error codes like 404
+            .map(|x| {
+                // TODO avoid copying somehow ?
+                x.as_string().expect("Error when converting into string")
+            })
+    }
 
-// todo fetch_text?
-
-/// Allow access to a json object, inside a callback.
-pub fn fetch_json(
-    method: Method,
-    url: &str,
-    request_opts: Option<RequestOpts>,
-    callback: Box<Fn(JsValue)>
-) {
-    let request = fetch_helper(method, url, request_opts);
-    let window = web_sys::window().expect("Can't find window");
-
-    // We could get the fetch_request_with_init promise directly from the func,
-    // but am running into issues with futures etc. So, repeated code here between
-    // this and fetch.
-    let request_promise = window.fetch_with_request(&request);
-    let f = wasm_bindgen_futures::JsFuture::from(request_promise)
-
-            .and_then(move |resp_value| {
-        // `resp_value` is a `Response` object.
-            assert!(resp_value.is_instance_of::<web_sys::Response>());
-            let resp: web_sys::Response = resp_value.dyn_into()
-                .expect("Problem casting response as Reponse.");
-
-            resp.json()
-//          resp.text()
-        })
-        .and_then(|json_value| {
-            // Convert this other `Promise` into a rust `Future`.
-            wasm_bindgen_futures::JsFuture::from(json_value)
-        })
-        .and_then(move |v| {
-            callback(v);
-            future::ok(JsValue::null())
-        });
-
-    wasm_bindgen_futures::future_to_promise(f);
+    pub fn fetch_json<A: DeserializeOwned>(self) -> impl Future<Item = A, Error = JsValue> {
+        self.fetch_string()
+            .map(|text| {
+                serde_json::from_str(&text).expect("Error deserializing JSON")
```

Comment:
Not sure what the error-handling story should be with this API, but this `expect` in particular would seem to panic if the server returns malformed data, which is not acceptable. I think instead the error should be propagated inside the `Future`

Reply:
I agree in principle (I'm hugely ***not*** a fan of `unwrap`/`expect`), however `JsFuture` has an error type of `JsValue`, whereas Serde does not. So it'll require some thinking on how to convert between them.